agent/connect_auth.go

```diff
@@ -0,0 +1,146 @@
+package agent
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/cache"
+	cachetype "github.com/hashicorp/consul/agent/cache-types"
+	"github.com/hashicorp/consul/agent/connect"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// ConnectAuthorize implements the core authorization logic for Connect. It's in
+// a separate agent method here because we need to re-use this both in our own
+// HTTP API authz endpoint and in the gRPX xDS/ext_authz API for envoy.
+//
+// The ACL token and the auth request are provided and the auth decision (true
+// means authorised) and reason string are returned.
+//
+// If the request input is invalide the error returned will be a
+// BadRequestError, if the token doesn't grant necessary access then an
+// acl.ErrPermissionDenied error is returned, otherwise error indicates an
+// unexpected server failure. If access is denied, no error is returned but the
+// first return value is false.
+func (a *Agent) ConnectAuthorize(token string,
+	req *structs.ConnectAuthorizeRequest) (authz bool, reason string, m *cache.ResultMeta, err error) {
+	if req == nil {
+		err = BadRequestError{"Invalid request"}
+		return
+	}
+
+	// We need to have a target to check intentions
+	if req.Target == "" {
+		err = BadRequestError{"Target service must be specified"}
+		return
+	}
+
+	// Parse the certificate URI from the client ID
+	uri, err := connect.ParseCertURIFromString(req.ClientCertURI)
```

Comment:
first place `err` is shadowed and from here out `err =...` isn't assigning to the named return parameters anymore sadly.

Reply:
This sort of failure mode of the function I think does bad things to the authorization.

Reply:
For future reference this is not shadowing `err` which is why tests pass even those asserting errors beyond this point. According to https://golang.org/ref/spec#Short_variable_declarations short declarations (`:=`) don't redeclare variables, they just assign the existing one the value without a compile error if it's same type and in scope and at least one of the names on the left hand side is new.

It's also a compile error if you shadow a return variable and then return implicitly: https://play.golang.org/p/QnUiepescSs

That said, I agree this is harder to read and more prone to subtleties with named returns so I intend to change it, good call!